stub/src/main/java/io/grpc/stub/ServerCalls.java

```diff
@@ -122,24 +129,35 @@ private ServerCalls() {
           ReqT request;
           @Override
           public void onMessage(ReqT request) {
+            if (this.request != null) {
+              // Safe to close the call, because the application has not yet been invoked
```

Comment:
Added the onHalfClose to the test and verified that this now passes the test with the boolean. Also realized that the server streaming tests should really be using the `ServerCalls.asyncServerStreamingCall` stub, even though it's equivalent to `asyncUnaryRequestCall`. I broke the tests into independent tests to be more precise here.